lls/cache.c

```diff
@@ -0,0 +1,481 @@
+			if (lls_conf->debug)
+				lls_cache_dump(cache);
+		}
+		return;
+	}
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: unrecognized request type (%d)\n", req->ty);
+		return;
+	}
+}
+
+unsigned int
+lls_cache_writer(struct lls_config *lls_conf)
+{
+	struct lls_request *reqs[LLS_CACHE_BURST_SIZE];
+	unsigned int count;
+	unsigned int i;
+
+	count = mb_dequeue_burst(&lls_conf->requests,
+		(void **)reqs, LLS_CACHE_BURST_SIZE);
+
+	if (likely(count == 0))
+		return count;
+
+	for (i = 0; i < count; i++) {
+		process_req(lls_conf, reqs[i]);
+		mb_free_entry(&lls_conf->requests, reqs[i]);
+	}
+
+	return count;
+}
+
+int
+lls_cache_request(struct lls_cache *cache, uint8_t *ip_be,
+	enum lls_req_ty ty, void *req_arg)
+{
+	struct lls_config *lls_conf = get_lls_conf();
+	struct lls_request *req = mb_alloc_entry(&lls_conf->requests);
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+	int ret;
+
+	if (req == NULL) {
+		ip_str = cache->ip_str(cache, ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: allocation for request of type %d for %s failed",
+			ty, ip_str == NULL ? "<unknown>" : ip_str);
+	}
+
+	req->ty = ty;
+	req->cache = cache;
+	memcpy(req->ip_be, ip_be, cache->key_len);
+
+	switch (ty) {
+	case LLS_REQ_RESOLVE:
+		/* FALLTHROUGH */
+	case LLS_REQ_SUBSCRIBE:
+		req->u.hold = *(struct lls_hold *)req_arg;
+		break;
+	case LLS_REQ_PUT:
+		req->u.put = *(struct lls_put *)req_arg;
+		break;
+	case LLS_REQ_MODIFY:
+		req->u.mod = *(struct lls_mod *)req_arg;
+		break;
+	default:
+		mb_free_entry(&lls_conf->requests, req);
+		ip_str = cache->ip_str(cache, ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: unknown request type %d for %s failed",
+			ty, ip_str == NULL ? "<unknown>" : ip_str);
+		return -1;
+	}
+
+	ret = mb_send_entry(&lls_conf->requests, req);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static void
+lls_cache_del(struct lls_cache *cache, struct lls_record *record)
+{
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+
+	int32_t ret = rte_hash_del_key(cache->hash, record->map.ip_be);
+	if (unlikely(ret == -ENOENT)) {
+		ip_str = cache->ip_str(cache, record->map.ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"No map found, record for %s not deleted\n",
+			ip_str == NULL ? cache->name : ip_str);
+			return;
+	} else if (unlikely(ret == -EINVAL)) {
+		ip_str = cache->ip_str(cache, record->map.ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, record for %s not deleted\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+}
+
+void
+lls_cache_scan(struct lls_cache *cache, struct gatekeeper_if *front,
+	struct gatekeeper_if *back)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	uint8_t *ip_be;
+	void *data;
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+	struct lls_config *lls_conf = get_lls_conf();
+	time_t now = time(NULL);
+
+	index = rte_hash_iterate(cache->hash, (void *)&ip_be, &data, &iter);
```

Comment:
I recommend to add `RTE_ASSERT(now >= 0);` before this line since `time()` can fail according to the manual! Please search the code for all calls to `time()` to do the same.